Prepare Pica registers for Geometry Shaders

JayFoxRox opened the pull request:
This code is part of the geometry shader (GS) implementation written by me which is based on work by @ds84182 .

This PR shouldn't change any behaviour currently (except maybe for cases where GS is active, but those are bugged anyway). It merely prepares the registers for GS.
I'd recommend inclusion in bleeding-edge to figure out if there are any issues.
However, this went through a larger refactor shortly before creating this PR, so review / testing should be done first.

Review per commit; see commit description too.

---

### Notes:

* The limit for VS (Vertex Shaders) appears to be 512 program words (compared to 4096 for GS). We should probably introduce a template parameter or something to handle these differences between VS / GS structure. However, that's beyond the scope of this PR. I can create an issue for this.

* [`vsh_com_mode`](https://3dbrew.org/wiki/GPU/Internal_Registers#GPUREG_VSH_COM_MODE) will later (next PR) be indicating if GS will use the same shader data (uniform / program / swizzle data) as VS.
It is not yet known if this influences writes (the functions introduced in this commit) OR reads (in the interpreter and jit).
We also don't know wether the shader data is per shader mode (VS / GS) or per shader-unit (4 units, only one of which is able to run GS).
However, this should be easy to change in the future if the current implementation is wrong (handling it on read = not doing extra work in functions introduced by this PR).
*We need a couple of hardware tests to understand how exactly the shader data is uploaded. However, I vote for accepting this for now, and refactoring later (once hardware tests have been done).*

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/2695)
<!-- Reviewable:end -->

Subv commented:
```
Note that some of the registers will be unused for a while (until the next PR at least). Shall I mark them with // Unused or something?
```
IMO, no

Subv commented on src/video_core/command_processor.cpp:
```diff
@@ -330,24 +326,90 @@ static void WritePicaReg(u32 id, u32 value, u32 mask) {
+    case PICA_REG_INDEX_WORKAROUND(gs.int_uniforms[2], 0x283):
+    case PICA_REG_INDEX_WORKAROUND(gs.int_uniforms[3], 0x284): {
+        unsigned index = (id - PICA_REG_INDEX_WORKAROUND(gs.int_uniforms[0], 0x281));
+        auto values = regs.gs.int_uniforms[index];
```
How is this different from the "value" passed as parameter to this function?

JayFoxRox replied:
No idea. This was naively copied from the `vs.int_uniforms` case.

Subv commented on src/video_core/shader/shader.cpp:
```diff
@@ -109,6 +109,124 @@ void Shutdown() {
 #endif // ARCHITECTURE_x86_64
 }
 
+bool SharedGS() {
```
Can we give this a better name and some documentation?

Subv commented on src/video_core/shader/shader.cpp:
```diff
@@ -109,6 +109,124 @@ void Shutdown() {
+              values.x.Value(), values.y.Value(), values.z.Value(), values.w.Value());
+}
+
+void WriteUniformFloatSetupReg(bool gs, u32 value) {
```
Is this function really needed? afaik the WritePicaReg function will write the value anyways.

JayFoxRox replied:
Until some hours ago these functions did a write-through / copy to GS registers on VS writes.
Depending on how the hardware-tests turns out this will also be re-added later.

Subv commented on src/video_core/shader/shader.cpp:
```diff
@@ -109,6 +109,124 @@ void Shutdown() {
+    uniform_write_buffer[float_regs_counter++] = value;
+
+    // Uniforms are written in a packed format such that four float24 values are
+    // encoded in
```
what's with this break?

JayFoxRox replied:
good question.. I assume clang-format broke it. - I'll fix this

Subv commented on src/video_core/shader/shader.cpp:
```diff
@@ -109,6 +109,124 @@ void Shutdown() {
+    }
+}
+
+void WriteProgramCodeOffset(bool gs, u32 value) {
```
same as WriteUniformFloatSetupReg

j-selby commented:
> The limit for VS (Vertex Shaders) appears to be 512 program words. We should probably introduce a template parameter or something to handle these differences between VS / GS structure. However, that's beyond the scope of this PR. I can create an issue for this.

Imo, this should definitely be recorded. It might be better to add a `TODO: Handle difference between GS and VS`. That might be a good introduction project for someone as well, so yeah, create a issue.

> Another little annoyance with this commit is that the functions take a bool as first argument for false = VS / true = GS. Shall I change that to an enum class; something else; or leave as-is?

enum - `bool gs` really doesn't look great on paper, though that's the Java programmer in me speaking :P

JayFoxRox commented on src/video_core/regs_pipeline.h:
```diff
@@ -200,7 +200,14 @@ struct PipelineRegs {
+    enum class VSComMode : u32 {
+        Shared = 0,
+        Exclusive = 1
+    };
```
Clang format proposes `enum class VSComMode : u32 { Shared = 0, Exclusive = 1 };` but I don't really like that (it's less readable imho). I'm also thinking that I've seen code like this in the past (and even copied it from elsewhere in Citra).

What should I change it into?

yuriks replied:
Add a comma after the last item, which will preserve the multi-line formatting.

yuriks commented:
You should add your commit descriptions to the actual commit messages.

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[src/video_core/regs_pipeline.h, line 205 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjfc7FYmYkx97lCTAiI:-Kjfc7FZU-UfuxYQ7zit:b61d1vr) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/regs_pipeline.h#L205)):*
> ```objc
>     INSERT_PADDING_WORDS(1);
> 
>     enum class VSComMode : u32 {
> ```

What's a "com mode"? "Command mode"? Make this name clearer.

---

*[src/video_core/regs_pipeline.h, line 207 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfcPSq2WyUtey7Lb7s:-KjfcPSq2WyUtey7Lb7t:b-3fttpi) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/regs_pipeline.h#L207)):*
> ```objc
>     enum class VSComMode : u32 {
>         Shared = 0,
>         Exclusive = 1
> ```

Also document what each of these values means. This whole reg needs some documentation.

---

*[src/video_core/shader/shader.cpp, line 112 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kje9SuJ0eryT4JCV7Fb-r1-112:-KjfguDKnrg01tRfnX1l:b-v6sxld) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L112)):*
<details><summary><i>Previously, Subv (Sebastian Valle) wrote…</i></summary><blockquote>

Can we give this a better name and some documentation?
</blockquote></details>

This doesn't even seem to be used?

---

*[src/video_core/shader/shader.cpp, line 116 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfeYcg6igOSIzb_vGo:-KjfeYcg6igOSIzb_vGp:b8xh5n9) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L116)):*
> ```cpp
> }
> 
> void WriteUniformBoolReg(bool gs, u32 value) {
> ```

50000% makes this an enum and not a bool. The bool is pretty much impossible to read on the call-site.

Or even better: in most cases where you just use the parameter to decide between which state or regs struct to use, just pass those in as parameters.

---

*[src/video_core/shader/shader.cpp, line 119 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjffOLo50lKrDO7TOsV:-KjffOLo50lKrDO7TOsW:bpctyv3) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L119)):*
> ```cpp
>     auto& setup = gs ? g_state.gs : g_state.vs;
> 
>     ASSERT(setup.uniforms.b.size() == 16);
> ```

Shouldn't this be a static_assert?

---

*[src/video_core/shader/shader.cpp, line 134 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kje9SuJ0eryT4JCV7Fb-r1-134:-KjfflKKPnFl4dDaPEGF:b-oa3djr) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L134)):*
<details><summary><i>Previously, JayFoxRox (Jannik Vogel) wrote…</i></summary><blockquote>

Until some hours ago these functions did a write-through / copy to GS registers on VS writes.
Depending on how the hardware-tests turns out this will also be re-added later.
</blockquote></details>

I say leave this function out until there comes a time when it becomes necessary, since right now it's not actually saving any code or work.

---

*[src/video_core/shader/shader.cpp, line 159 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfgGSHCALC1U_l7rSl:-KjfgGSI0U4uhMq9lJ-z:b-lqgcvz) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L159)):*
> ```cpp
>     // written.
>     if ((float_regs_counter >= 4 && uniform_setup.IsFloat32()) ||
>         (float_regs_counter >= 3 && !uniform_setup.IsFloat32())) {
> ```

`if (float_regs_counter >= (uniform_setup.IsFloat32() ? 4 : 3)) {`, maybe make the latter part a separate variable.

---

*[src/video_core/shader/shader.cpp, line 211 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjfh4A9doctW9mdxomv:-Kjfh4AAZneXiRzwsxf1:bw47sfw) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L211)):*
> ```cpp
> }
> 
> void WriteSwizzlePatternsOffset(bool gs, u32 value) {
> ```

Same as the other trivial funcs pointed out by Subv

---

*[src/video_core/shader/shader.h, line 148 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfiqORCxfRmxK0MzHU:-KjfXm92V2viUXhEaIU_:b-4w0xof) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.h#L148)):*
> ```objc
> 
>     int float_regs_counter = 0;
>     u32 uniform_write_buffer[4];
> ```

This doesn't seem like a good place to put these. Have you verified that both units have a separate counter/buffer for this? If not, then this seems like a premature change to make.

---

*[src/video_core/shader/shader.h, line 192 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjfj9dl3_CQdjYrBKo_:-Kjfj9dl3_CQdjYrBKoa:b-akp0ct) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.h#L192)):*
> ```objc
> void WriteProgramCode(bool gs, u32 value);
> void WriteSwizzlePatternsOffset(bool gs, u32 value);
> void WriteSwizzlePatterns(bool gs, u32 value);
> ```

IMO these should go into command_processor.cpp, or be split between that (incoming value parsing) and member vars in the relevant structs (applying the value, iff it's not trivial)

---

*[src/video_core/shader/shader_jit_x64_compiler.h, line 26 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfhcJeKCKx996eW3ji:-KjfhcJeKCKx996eW3jj:b15xzek) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader_jit_x64_compiler.h#L26)):*
> ```objc
> 
> /// Memory allocated for each compiled shader (256Kb)
> constexpr size_t MAX_SHADER_SIZE = 1024 * 256;
> ```

Probably `4096 * 64` makes more sense? Also, maybe you should introduce a constant for the maximum shader size so it's just one place to change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695#-:-KjfkkoC35nXey3I_tW6:bv3833y)*
<!-- Sent from Reviewable.io -->

JayFoxRox commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[src/video_core/regs_pipeline.h, line 205 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjfc7FYmYkx97lCTAiI:-Kjg22vIGVvfjau4r-mg:b-vv7tla) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/regs_pipeline.h#L205)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

What's a "com mode"? "Command mode"? Make this name clearer.
</blockquote></details>

This is the name from https://3dbrew.org/wiki/GPU/Internal_Registers#GPUREG_VSH_COM_MODE .

The existing documentation is very vague about this and I'm not sure what the com stands for myself tbh..

---

*[src/video_core/shader/shader.cpp, line 112 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kje9SuJ0eryT4JCV7Fb-r1-112:-Kjg2YsTc9_eIGuKMeDg:bwbp4bg) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L112)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

This doesn't even seem to be used?
</blockquote></details>

I can remove `VSComMode` and `SharedGS()` temporarily and add them later as they are currently unused.

---

*[src/video_core/shader/shader.cpp, line 116 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfeYcg6igOSIzb_vGo:-Kjg2sTxHQ8v6Mq_BCxw:b-g92sez) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L116)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

50000% makes this an enum and not a bool. The bool is pretty much impossible to read on the call-site.

Or even better: in most cases where you just use the parameter to decide between which state or regs struct to use, just pass those in as parameters.
</blockquote></details>

- Will do.
- Will do.

---

*[src/video_core/shader/shader.cpp, line 119 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjffOLo50lKrDO7TOsV:-Kjg33_42O3wic9qK0Gr:b-i2ii) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L119)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Shouldn't this be a static_assert?
</blockquote></details>

Probably.

---

*[src/video_core/shader/shader.cpp, line 159 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfgGSHCALC1U_l7rSl:-Kjg3FGIBEn4Obdq6ynn:b5p0jno) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L159)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

`if (float_regs_counter >= (uniform_setup.IsFloat32() ? 4 : 3)) {`, maybe make the latter part a separate variable.
</blockquote></details>

This is existing code which was just moved.
I don't want to refactor too much at this point.

---

*[src/video_core/shader/shader.h, line 148 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfiqORCxfRmxK0MzHU:-Kjg3ZnEwyirVt3GHtEc:b4lrts3) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.h#L148)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

This doesn't seem like a good place to put these. Have you verified that both units have a separate counter/buffer for this? If not, then this seems like a premature change to make.
</blockquote></details>

Agreed. Again: Just moved existing code. The place to put them is not ideal.
There is a TODO in `WriteUniformFloatReg()`.
I can create an issue about this after merge?

---

*[src/video_core/shader/shader.h, line 192 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjfj9dl3_CQdjYrBKo_:-Kjg4_Q0XtALxUv8TlB5:bdbsyfq) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.h#L192)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

IMO these should go into command_processor.cpp, or be split between that (incoming value parsing) and member vars in the relevant structs (applying the value, iff it's not trivial)
</blockquote></details>

This is more or less a temporary solution until we know wether there is some write-through from VS to GS and wether this information is stored per shader-unit or shader type.
(At which point we can move them into the appropriate objects)

I'm not sure how you'd want to split this.

---

*[src/video_core/shader/shader_jit_x64_compiler.h, line 26 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfhcJeKCKx996eW3ji:-Kjg42vvh19V7AaTZI-0:brkw54x) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader_jit_x64_compiler.h#L26)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Probably `4096 * 64` makes more sense? Also, maybe you should introduce a constant for the maximum shader size so it's just one place to change.
</blockquote></details>

Yeah, I just wasn't sure wether the 1024 was used to get the value in Kb or it was being used to represent the number of instructions.
It was more or less a quickfix as the current size will result in xbyak running out of memory.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

yuriks commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[src/video_core/shader/shader.h, line 148 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfiqORCxfRmxK0MzHU:-Kjida2QQhecJM3vdUVN:b-e7a6rx) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.h#L148)):*
<details><summary><i>Previously, JayFoxRox (Jannik Vogel) wrote…</i></summary><blockquote>

Agreed. Again: Just moved existing code. The place to put them is not ideal.
There is a TODO in `WriteUniformFloatReg()`.
I can create an issue about this after merge?
</blockquote></details>

You moved this code yourself as a separate change, it's not related to the moved code below? I think it doesn't make sense to make this part of the struct. Use separate static counters instead if necessary. These are more part of the command interface, not the shader engine state.

---

*[src/video_core/shader/shader.h, line 192 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjfj9dl3_CQdjYrBKo_:-KjihQvYn6uzWU3BvEiV:bua1pv7) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.h#L192)):*
<details><summary><i>Previously, JayFoxRox (Jannik Vogel) wrote…</i></summary><blockquote>

This is more or less a temporary solution until we know wether there is some write-through from VS to GS and wether this information is stored per shader-unit or shader type.
(At which point we can move them into the appropriate objects)

I'm not sure how you'd want to split this.
</blockquote></details>

As mentioned in IRC, I think they're better off in command_processor.cpp

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

JayFoxRox commented:
Commit descriptions have been added, the first post has been converted into purely notes.

I've tried to make the requested changes.
I've decided to drop one of the commits - it will be added to the next PR when the introduced variables will actually be used.

---

Review status: 3 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/video_core/shader/shader.h, line 148 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfiqORCxfRmxK0MzHU:-KjiwGjInjcLsgRSMW7d:b-85oa8p) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.h#L148)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

You moved this code yourself as a separate change, it's not related to the moved code below? I think it doesn't make sense to make this part of the struct. Use separate static counters instead if necessary. These are more part of the command interface, not the shader engine state.
</blockquote></details>

Moved back to where it was.
I've introduced a seperate static buffer.

---

*[src/video_core/regs_pipeline.h, line 207 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjfcPSq2WyUtey7Lb7s:-KjivMsaQbjMIJkZ99xB:b-4hzvdi) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/regs_pipeline.h#L207)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Also document what each of these values means. This whole reg needs some documentation.
</blockquote></details>

I'll remove this commit from the PR.
It's only necessary when implementing the actual GS logic (it was necessary until like 10 minutes before I created the PR, when SharedGS() etc were still being used in Write*() functions).

---

*[src/video_core/shader/shader.cpp, line 112 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kje9SuJ0eryT4JCV7Fb-r1-112:-Kjivlj-lGbp0Ik-i9YR:b-wssd43) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L112)):*
<details><summary><i>Previously, JayFoxRox (Jannik Vogel) wrote…</i></summary><blockquote>

I can remove `VSComMode` and `SharedGS()` temporarily and add them later as they are currently unused.
</blockquote></details>

Dropped this commit altogether.

---

*[src/video_core/shader/shader.cpp, line 134 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kje9SuJ0eryT4JCV7Fb-r1-134:-KjivrTIk1TWEAl3Ln-W:bbox8rh) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L134)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

I say leave this function out until there comes a time when it becomes necessary, since right now it's not actually saving any code or work.
</blockquote></details>

Function dropped

---

*[src/video_core/shader/shader.cpp, line 211 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjfh4A9doctW9mdxomv:-KjiympWJa0L4eKqK3sw:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L211)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Same as the other trivial funcs pointed out by Subv
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695#-:-Kjivckak-1-f0H2pL1Z:b-6nac5)*
<!-- Sent from Reviewable.io -->

yuriks commented:
Reviewed 2 of 7 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/video_core/shader/shader.h, line 147 at r4](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjkSY9YAWGaDHTMutvI:-KjkSY9YAWGaDHTMutvJ:b-6ijrhi) ([raw file](https://github.com/citra-emu/citra/blob/7cb79e3e9d04874ba5ec33b2c48173831c55d60b/src/video_core/shader/shader.h#L147)):*
> ```objc
>     }
> 
>     std::array<u32, 4096> program_code;
> ```

I think it's worth taking the opportunity to make a separate named constant for the maximum shader size, since you're already touching all places where it's repeated in this commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

JayFoxRox commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/video_core/shader/shader.h, line 147 at r4](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjkSY9YAWGaDHTMutvI:-Kjn6bty9GpWVTbLEEe4:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/7cb79e3e9d04874ba5ec33b2c48173831c55d60b/src/video_core/shader/shader.h#L147)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

I think it's worth taking the opportunity to make a separate named constant for the maximum shader size, since you're already touching all places where it's repeated in this commit.
</blockquote></details>

Done.

---

*[src/video_core/shader/shader.cpp, line 193 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kje9SuJ0eryT4JCV7Fb-r1-193:-Kjn6fuLyZiA0csqHmc-:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/f0577160510c043f9f3ea2a4dad7b7adcdcc54e8/src/video_core/shader/shader.cpp#L193)):*
<details><summary><i>Previously, Subv (Sebastian Valle) wrote…</i></summary><blockquote>

same as WriteUniformFloatSetupReg
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

yuriks commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 7 files at r2, 6 of 6 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/video_core/shader/shader.h, line 28 at r5](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjp8tAGySYtyBTiP-7P:-Kjp8tAHJxoXNnz0AgR0:b-qc5h91) ([raw file](https://github.com/citra-emu/citra/blob/9500b5d9f6472c4953437fbd9ce47d135131b22b/src/video_core/shader/shader.h#L28)):*
> ```objc
> 
> constexpr unsigned MaxProgramCodeLength = 4096;
> constexpr unsigned MaxSwizzleDataLength = 4096;
> ```

These should be either `max_swizzle_data_length` or `MAX_SWIZZLE_DATA_LENGTH` (we're inconsistent on which of those two we use, but CamelCase is reserved for functions and classes.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695#-:-KjpA2fHfRvmaZWg5e8f:bnfp4nl)*
<!-- Sent from Reviewable.io -->

yuriks commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjpAzUbWiBvy14wEcjx:-KjpAzUbWiBvy14wEcjy:b-qr1kvt) (no related file):*
Both builds are failing, you should look into that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

wwylele commented on src/video_core/command_processor.cpp:
```diff
@@ -48,6 +49,99 @@ static const u32 expand_bits_to_bytes[] = {
+}
+
+static void WriteUniformBoolReg(Shader::ShaderSetup& setup, u32 value) {
+    static_assert(setup.uniforms.b.size() == 16);
```
`static_assert` needs a string as an error message (the no-message version is only in c++17 iirc)

wwylele commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/video_core/command_processor.cpp, line 63 at r7](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjq1u9_frQiGZmHBcu-:-Kjq1u9_frQiGZmHBcu0:b-yx3o1r) ([raw file](https://github.com/citra-emu/citra/blob/5d82c9aa63276b0ddd3b785b6c13cbbe562affa0/src/video_core/command_processor.cpp#L63)):*
> ```cpp
> 
> static void WriteUniformBoolReg(Shader::ShaderSetup& setup, u32 value) {
>     static_assert(setup.uniforms.b.size() == 16);
> ```

`static_assert` needs a string as an error message (the no-message version is only in c++17 iirc)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

JayFoxRox commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjpAzUbWiBvy14wEcjx:-KjrbCCcmqVGlpjbrz8Y:b-q3pqso) (no related file):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Both builds are failing, you should look into that.
</blockquote></details>

Builds failed because of the `static_assert`

---

*[src/video_core/command_processor.cpp, line 63 at r7](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjq1u9_frQiGZmHBcu-:-Kjr_81tTRvpp5ELRIjS:b-2j2mt3) ([raw file](https://github.com/citra-emu/citra/blob/5d82c9aa63276b0ddd3b785b6c13cbbe562affa0/src/video_core/command_processor.cpp#L63)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

`static_assert` needs a string as an error message (the no-message version is only in c++17 iirc)
</blockquote></details>

Dropped the assert altogether (it used to be `ASSERT()`, then `static_assert`, but the MacOS version didn't like that [with or without strings])

---

*[src/video_core/shader/shader.h, line 28 at r5](https://reviewable.io:443/reviews/citra-emu/citra/2695#-Kjp8tAGySYtyBTiP-7P:-KjraRJ4m_FcbVUpmpAe:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/9500b5d9f6472c4953437fbd9ce47d135131b22b/src/video_core/shader/shader.h#L28)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

These should be either `max_swizzle_data_length` or `MAX_SWIZZLE_DATA_LENGTH` (we're inconsistent on which of those two we use, but CamelCase is reserved for functions and classes.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

yuriks commented:
Reviewed 5 of 6 files at r8, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

wwylele commented:
Mostly LGTM. Please make sure all `LOG_TRACE` still works

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/video_core/command_processor.cpp, line 72 at r10](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjsQZtTqxhJUbFuHS03:-KjsQZtTqxhJUbFuHS04:b-5805au) ([raw file](https://github.com/citra-emu/citra/blob/8af6ba4a86611675b9020f2780d0fb68f6f903e4/src/video_core/command_processor.cpp#L72)):*
> ```cpp
>     setup.uniforms.i[index] = values;
>     LOG_TRACE(HW_GPU, "Set %s integer uniform %d to %02x %02x %02x %02x",
>               GetShaderSetupTypeName(setup), index, values.x.Value(), values.y.Value(),
> ```

Remove `Value()`. You should also check if debug build also compile.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695#-:-KjsQwavIr9-7YpocpUs:b-z5t7jx)*
<!-- Sent from Reviewable.io -->

Subv commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695#-:-Kjua2-WAZE9n5okrgsm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

JayFoxRox commented:
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/video_core/command_processor.cpp, line 72 at r10](https://reviewable.io:443/reviews/citra-emu/citra/2695#-KjsQZtTqxhJUbFuHS03:-Kjx2InklAIcAOx6jLos:bt43niw) ([raw file](https://github.com/citra-emu/citra/blob/8af6ba4a86611675b9020f2780d0fb68f6f903e4/src/video_core/command_processor.cpp#L72)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

Remove `Value()`. You should also check if debug build also compile.
</blockquote></details>

Done.

Compiled fine with `-DCMAKE_BUILD_TYPE=Debug`.
MK logs ints "01 00 00 00" to trace at startup (as it does on master), so the log seems to work fine too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695)*
<!-- Sent from Reviewable.io -->

wwylele commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> will merge after CI check

---

Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2695#-:-Kjx50cB1-st6mW4s70E:b-o50knb)*
<!-- Sent from Reviewable.io -->
